[HW] Add high(er) level module builder

mortbopet opened the pull request:
Picking up some of the capabilities presented in #3781 but with a builder-based approach. One of the biggest "wants" from #3781 (for me, personally) is to have a way of building modules where i can easily look up input ports and assign to output ports.

Example use:
```c++
auto modports = hw::ModulePortInfo( ... ); // (in0: i32, in1: i32) -> (out: i32)
auto module = b.create<hw::HWModuleOp>(loc, b.getStringAttr("foo"), modports,
    [&](OpBuilder&b, HWModulePortAccessor& ports) {
  auto res = b.create<comb::AddOp>(loc, ports["in0"], ports["in1"]);
  ports["out0"].assign(res);
});
```

No test included with this PR - didn't find it warranted to create a separate test pass just for this (and i plan to use it in an upcoming commit).

teqdruid commented:
Is it desirable at this point to create backedges? I suspect that just providing a `set("outputPortName", value)` method would be sufficient and avoid a bunch of complexity. Users could always create the backedges themselves if necessary. This is what we do in pycde. Would that add significant burden on the code you want to use this for?

In lieu of a test, could you port some existing code to use this API? Or is there nothing currently existing which makes sense for this API?

mortbopet commented:
> Is it desirable at this point to create backedges? 

I do think it brings a non-negligable benefit of being able to inspect the `hw.output` operands without worrying about having to create ones' own backedges.

>  Users could always create the backedges themselves if necessary. This is what we do in pycde.

So, as far as i can tell, the PyCDE way of doing these port module accessors is to [record output operand values separately](https://github.com/llvm/circt/blob/main/frontends/PyCDE/src/pycde/module.py#L603) and then later fixup the `hw.output` operation to contain these operands. Personally, I think backedges are a better way to do this, since those two approaches essentially does the same - backedges just encapsulate that concept for a single operand. + (as mentioned above) with the added benefit that it is legal to access the `hw.output` operand at any point in time.

> In lieu of a test, could you port some existing code to use this API? Or is there nothing currently existing which makes sense for this API?

I haven't looked at the entire codebase but i could give it a second scan. Was considering the FSM lowering, however there, building the internals of the `hw.module` far outweighs the LoC spent on defining the I/O of that module.

mortbopet commented:
Changed implementation to maintaining a record of output operand values, and removed use of backedges - after trying it out, I agree that it's a better (re: more maintainable) solution.

Discussed implementation offline with @teqdruid hence why I'm checking in this as-is to unblock me. Please leave a follow-up comment if there's anything you'd like to see changed.

teqdruid commented on include/circt/Dialect/HW/HWOps.h:
```diff
@@ -25,6 +25,8 @@
 #include "mlir/Interfaces/SideEffectInterfaces.h"
 #include "llvm/ADT/StringExtras.h"
 
+#include <map>
```
Does `llvm::StringMap` not work for some reason? (If you haven't seen it, I use the [LLVM Programmer's Manual(https://www.llvm.org/docs/ProgrammersManual.html) for a guide to LLVM datastructures.)

teqdruid commented on include/circt/Support/BackedgeBuilder.h:
```diff
@@ -89,6 +89,7 @@ class Backedge {
   explicit operator bool() const { return !!value; }
   operator mlir::Value() const { return value; }
   void setValue(mlir::Value);
+  bool isSet() { return set; }
```
Since you're not using backedges anymore, is this used?

teqdruid commented on include/circt/Dialect/HW/HWOps.h:
```diff
@@ -188,6 +190,38 @@ StringAttr getArgSym(Operation *op, unsigned i);
+
+private:
+  std::map<std::string, unsigned> inputIdx, outputIdx;
+  DenseMap<unsigned, Value> inputArgs;
```
Is a `DenseMap` required? Why not use a `SmallVector` given that ports _should_ fill in all the slots?

teqdruid commented on lib/Dialect/HW/HWOps.cpp:
```diff
@@ -178,6 +178,28 @@ StringAttr hw::getResultSym(Operation *op, unsigned i) {
+    inputArgs[inputInfo.argNum] = barg;
+  }
+
+  llvm::SmallVector<Value> outputOpArgs;
```
This doesn't appear to be used anywhere.

teqdruid commented on lib/Dialect/HW/HWOps.cpp:
```diff
@@ -576,6 +598,29 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &result,
+  build(builder, odsState, name, ports, parameters, attributes, comment);
+  auto *bodyRegion = odsState.regions[0].get();
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(&bodyRegion->front());
```
I think this is effectively a no-op.

teqdruid commented on lib/Dialect/HW/HWOps.cpp:
```diff
@@ -576,6 +598,29 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &result,
+  // Replace output operands.
+  auto outputOp = cast<hw::OutputOp>(bodyRegion->front().getTerminator());
+  llvm::SmallVector<Value> outputOperands;
+  for (auto [idx, operand] : accessor.getOutputOperands()) {
```
Order is not guaranteed to be `idx` sorted since you're using a `DenseMap`.

teqdruid replied:
It's not even guaranteed to be deterministic.

teqdruid reviewed (commented):
Also missing a test, which should be moving one of the existing module builders over to use this API.

mortbopet commented:
Review comments addressed in #3844
